Pyjion/intrins.cpp

```diff
@@ -872,99 +873,89 @@ int PyJit_ImportStar(PyObject*from, PyFrameObject* f) {
-        // need to concat values together...
-        auto finalArgs = PyTuple_New(PyTuple_GET_SIZE(args) + nstar);
-        for (int i = 0; i < PyTuple_GET_SIZE(args); i++) {
-            auto item = PyTuple_GET_ITEM(args, i);
-            Py_INCREF(item);
-            PyTuple_SET_ITEM(finalArgs, i, item);
-        }
-        for (int i = 0; i < PyTuple_GET_SIZE(stararg); i++) {
-            auto item = PyTuple_GET_ITEM(stararg, i);
-            Py_INCREF(item);
-            PyTuple_SET_ITEM(
-                finalArgs,
-                i + PyTuple_GET_SIZE(args),
-                item
-                );
-        }
-        Py_DECREF(args);
-        args = finalArgs;
-    }
-
-    result = PyObject_Call(func, args, kwargs);
-
-ext_call_fail:
-    Py_XDECREF(func);
-    Py_XDECREF(args);
-    Py_XDECREF(kwargs);
-    Py_XDECREF(stararg);
-    Py_XDECREF(kwdict);
-    return result;
+PyObject* PyJit_CallKwArgs(PyObject* func, PyObject*callargs, PyObject*kwargs) {
+	PyObject* result = nullptr;
+	if (!PyDict_CheckExact(kwargs)) {
+		PyObject *d = PyDict_New();
+		if (d == NULL)
+			goto error;
+		if (PyDict_Update(d, kwargs) != 0) {
+			Py_DECREF(d);
+			/* PyDict_Update raises attribute
+			* error (percolated from an attempt
+			* to get 'keys' attribute) instead of
+			* a type error if its second argument
+			* is not a mapping.
+			*/
+			if (PyErr_ExceptionMatches(PyExc_AttributeError)) {
+				PyErr_Format(PyExc_TypeError,
+					"%.200s%.200s argument after ** "
+					"must be a mapping, not %.200s",
+					PyEval_GetFuncName(func),
+					PyEval_GetFuncDesc(func),
+					kwargs->ob_type->tp_name);
+			}
+			goto error;
+		}
+		Py_DECREF(kwargs);
+		kwargs = d;
+	}
+
+	if (!PyTuple_CheckExact(callargs)) {
+		if (Py_TYPE(callargs)->tp_iter == NULL &&
+			!PySequence_Check(callargs)) {
+			PyErr_Format(PyExc_TypeError,
+				"%.200s%.200s argument after * "
+				"must be an iterable, not %.200s",
+				PyEval_GetFuncName(func),
+				PyEval_GetFuncDesc(func),
+				callargs->ob_type->tp_name);
+			goto error;
+		}
+
+		auto tmp = PySequence_Tuple(callargs);
+		if (tmp == nullptr) {
+			goto error;
+		}
+		Py_DECREF(callargs);
+		callargs = tmp;
+	}
+
+	result = PyObject_Call(func, callargs, kwargs);
+error:
+	Py_DECREF(func);
+	Py_DECREF(callargs);
+	Py_DECREF(kwargs);
+	return result;
+}
+
+PyObject* PyJit_CallArgs(PyObject* func, PyObject*callargs) {
+	PyObject* result = nullptr;
+	if (!PyTuple_CheckExact(callargs)) {
+		if (Py_TYPE(callargs)->tp_iter == NULL &&
+			!PySequence_Check(callargs)) {
+			PyErr_Format(PyExc_TypeError,
+				"%.200s%.200s argument after * "
+				"must be an iterable, not %.200s",
+				PyEval_GetFuncName(func),
+				PyEval_GetFuncDesc(func),
+				callargs->ob_type->tp_name);
+			goto error;
+		}
+		auto tmp = PySequence_Tuple(callargs);
+		if (tmp == nullptr) {
+			goto error;
+			return nullptr;
+		}
+		Py_DECREF(callargs);
+		callargs = tmp;
+	}
+
+	result = PyObject_Call(func, callargs, nullptr);
```

Comment:
I assume the tricks that Victor used for `CALL_METHOD` and such to avoid constructing the tuple for the call?